service/tablet_allocator.cc

```diff
@@ -1149,6 +1150,58 @@ class load_balancer {
     }
 };
 
+future<tablet_reallocation_result>
+reallocate_tablets_for_new_rf(const tablet_aware_replication_strategy* rep, schema_ptr s, token_metadata_ptr tm,
+    std::unordered_map<sstring, sstring> const& ks_options) {
+
+    auto* net_rep = dynamic_cast<const locator::network_topology_strategy*>(rep);
+
+    auto table_id = s->id();
+    tablet_map old_tablets = tm->tablets().get_tablet_map(table_id);
+    auto make_statuses = [&](tablet_reallocation_status status){
+        tablet_reallocation_result::status_map operation_statuses;
+        for (const auto& [dc, _] : ks_options) {
+            operation_statuses[dc].insert(status);
+        }
+        return operation_statuses;
+    };
+
+    if (!net_rep) {
+        co_return tablet_reallocation_result{ old_tablets, make_statuses(tablet_reallocation_status::unknown_topology_strategy) };
+    }
+
+    std::map<sstring, sstring> map_params(ks_options.cbegin(), ks_options.cend());
+    auto tablet_count = old_tablets.tablet_count();
+    locator::replication_strategy_params params{map_params, tablet_count};
+
+    auto new_rep = abstract_replication_strategy::create_replication_strategy(
+                "NetworkTopologyStrategy", params);
+
+    auto tablet_aware = new_rep->maybe_as_tablet_aware();
+
+    if (!tablet_aware) {
+        co_return tablet_reallocation_result{ old_tablets, make_statuses(tablet_reallocation_status::unknown_topology_strategy) };
+    }
+
+    try {
+        auto new_tablets = co_await tablet_aware->reallocate_tablets(s, tm, old_tablets);
+        co_return tablet_reallocation_result{ new_tablets, make_statuses(tablet_reallocation_status::success) };
+    } catch (exceptions::configuration_exception const& e) {
+        lblogger.warn("reallocate_tablets_for_new_rf_new: {}", e.what());
+        thread_local boost::regex re("Datacenter [0-9]+ doesn't have enough nodes for replication_factor=[0-9]+");
+        boost::cmatch what;
+        if (boost::regex_search(e.what(), what, re)) {
```

Comment:
we shouldn't rely on test searching error messages internally in scylla.
If it's important to treat this as a special case, I suggest we formally expose `check_enough_endpoints` from `network_topology_strategy` and call it here.